plugin/auth_ldap/src/auth_ldap_connection.cc

```diff
@@ -0,0 +1,288 @@
+  std::stringstream str_stream;
+  str_stream << (ssl ? "ldaps://" : "ldap://") << server_host << ":"
+             << server_port;
+  return str_stream.str();
+}
+
+bool AuthLDAPConnection::initiate(std::string server_host,
+                                  unsigned int server_port, bool ssl, bool tls,
+                                  std::string ca_path) {
+  this->created_ts = std::time(nullptr);
+  this->borrowed = false;
+
+  if (server_host.empty()) {
+    // log_srv_error("ERROR: server_host is empty");
+    return false;
+  }
+
+  int err = ldap_initialize(
+      &(this->ldap), get_ldap_uri(server_host, server_port, ssl).c_str());
+  if (err != LDAP_SUCCESS) {
+    std::stringstream log_stream;
+    log_stream << "ERROR: ldap_initialize " << ldap_err2string(err);
+    log_srv_error(log_stream.str());
+    return false;
+  }
+
+  int version = LDAP_VERSION3;
+  err = ldap_set_option(this->ldap, LDAP_OPT_PROTOCOL_VERSION, &version);
+  if (err != LDAP_OPT_SUCCESS) {
+    std::stringstream log_stream;
+    log_stream << "ERROR: ldap_set_option(LDAP_OPT_PROTOCOL_VERSION) "
+               << ldap_err2string(err);
+    log_srv_error(log_stream.str());
+    return false;
+  }
+
+  ldap_set_option(this->ldap, LDAP_OPT_REFERRALS, LDAP_OPT_OFF);
+
+  ldap_set_option(this->ldap, LDAP_OPT_RESTART, LDAP_OPT_ON);
+
+  if (ca_path.size() == 0) {
+    int reqCert = LDAP_OPT_X_TLS_NEVER;
+    err = ldap_set_option(this->ldap, LDAP_OPT_X_TLS_REQUIRE_CERT, &reqCert);
+    if (err != LDAP_OPT_SUCCESS) {
+      std::stringstream log_stream;
+      log_stream << "ERROR: ldap_set_option(LDAP_OPT_X_TLS_REQUIRE_CERT) "
+                 << ldap_err2string(err);
+      log_srv_error(log_stream.str());
+      return false;
+    }
+  } else {
+    char *cca_path = strdup(ca_path.c_str());
+    err = ldap_set_option(this->ldap, LDAP_OPT_X_TLS_CACERTFILE,
+                          static_cast<void *>(cca_path));
+    delete cca_path;
+    if (err != LDAP_OPT_SUCCESS) {
+      std::stringstream log_stream;
+      log_stream << "ERROR: ldap_set_option(LDAP_OPT_X_TLS_CACERTFILE) "
+                 << ldap_err2string(err);
+      log_srv_error(log_stream.str());
+      return false;
+    }
+  }
+
+  int new_ctx = 0;
+  err = ldap_set_option(this->ldap, LDAP_OPT_X_TLS_NEWCTX, &new_ctx);
+  if (err != LDAP_OPT_SUCCESS) {
+    std::stringstream log_stream;
+    log_stream << "ERROR: ldap_set_option(LDAP_OPT_X_TLS_NEWCTX) "
+               << ldap_err2string(err);
+    log_srv_error(log_stream.str());
+    return false;
+  }
+
+  if (tls) {
+    err = ldap_start_tls_s(this->ldap, nullptr, nullptr);
+    if (err != LDAP_SUCCESS) {
+      std::stringstream log_stream;
+      log_stream << "ERROR: ldap_start_tls_s " << ldap_err2string(err);
+      log_srv_error(log_stream.str());
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool AuthLDAPConnection::is_alive() {
+  int id;
+  return ldap_whoami(this->ldap, nullptr, nullptr, &id) == LDAP_SUCCESS;
```

Comment:
Can ldap_whoami handle NULLs? And similarly for other functions: shouldn't a null ldap connection result in an early exit?

Reply:
Yes, it can. Here we are checking that the connection is alive, since that would be the only case where ldap_whoami returns a non-success code

Reply:
Looking at the openldap github sources, there are some asserts about this:

https://github.com/openldap/openldap/blob/b06f5b0493937fc28f2cc86df1d7f464aa4504d8/libraries/libldap/whoami.c#L67

As those are asserts they won't fire in a release build with openldap, but I don't think that means it's supported.